File: components/cronet/android/test/cronet_test_util.cc
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/cronet/android/test/cronet_test_util.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "components/cronet/android/cronet_url_request_adapter.h"
#include "components/cronet/android/cronet_url_request_context_adapter.h"
#include "components/cronet/android/test/native_test_server.h"
#include "components/cronet/android/url_request_context_adapter.h"
#include "jni/CronetTestUtil_jni.h"
#include "net/dns/host_resolver_impl.h"
#include "net/dns/mock_host_resolver.h"
#include "net/url_request/url_request.h"

using base::android::JavaParamRef;

namespace cronet {

const char kFakeSdchDomain[] = "fake.sdch.domain";
// This must match the certificate used
// (quic_test.example.com.crt and quic_test.example.com.key.pkcs8), and
// the file served (
// components/cronet/android/test/assets/test/quic_data/simple.txt).
const char kFakeQuicDomain[] = "test.example.com";

namespace {

// Install host resolver rules to map fake domains to |destination|, usually an
// IP address.
void RegisterHostResolverProcHelper(net::URLRequestContext* url_request_context,
                                    const std::string& destination) {
  net::HostResolverImpl* resolver =
      static_cast<net::HostResolverImpl*>(url_request_context->host_resolver());
  scoped_refptr<net::RuleBasedHostResolverProc> proc =
      new net::RuleBasedHostResolverProc(NULL);
  proc->AddRule(kFakeSdchDomain, destination);
  proc->AddRule(kFakeQuicDomain, destination);
  resolver->set_proc_params_for_test(
      net::HostResolverImpl::ProcTaskParams(proc.get(), 1u));
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_CronetTestUtil_onHostResolverProcRegistered(env);
}

void RegisterHostResolverProcOnNetworkThread(
    CronetURLRequestContextAdapter* context_adapter,
    const std::string& destination) {
  RegisterHostResolverProcHelper(context_adapter->GetURLRequestContext(),
                                 destination);
}

// TODO(xunjieli): Delete this once legacy API is removed.
void RegisterHostResolverProcOnNetworkThreadLegacyAPI(
    URLRequestContextAdapter* context_adapter,
    const std::string& destination) {
  RegisterHostResolverProcHelper(context_adapter->GetURLRequestContext(),
                                 destination);
}

}  // namespace

void RegisterHostResolverProc(JNIEnv* env,
                              const JavaParamRef<jclass>& jcaller,
                              jlong jadapter,
                              jboolean jlegacy_api,
                              const JavaParamRef<jstring>& jdestination) {
  std::string destination(
      base::android::ConvertJavaStringToUTF8(env, jdestination));
  if (jlegacy_api == JNI_TRUE) {
    URLRequestContextAdapter* context_adapter =
        reinterpret_cast<URLRequestContextAdapter*>(jadapter);
    context_adapter->PostTaskToNetworkThread(
        FROM_HERE, base::Bind(&RegisterHostResolverProcOnNetworkThreadLegacyAPI,
                              base::Unretained(context_adapter), destination));
  } else {
    CronetURLRequestContextAdapter* context_adapter =
        reinterpret_cast<CronetURLRequestContextAdapter*>(jadapter);
    context_adapter->PostTaskToNetworkThread(
        FROM_HERE, base::Bind(&RegisterHostResolverProcOnNetworkThread,
                              base::Unretained(context_adapter), destination));
  }
}

jint GetLoadFlags(JNIEnv* env,
                  const JavaParamRef<jclass>& jcaller,
                  const jlong urlRequest) {
  return reinterpret_cast<CronetURLRequestAdapter*>(urlRequest)
      ->GetURLRequestForTesting()
      ->load_flags();
}

bool RegisterCronetTestUtil(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

}  // namespace cronet

File: components/cronet/android/test/src/org/chromium/net/CronetTestUtil.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.net;

import android.os.ConditionVariable;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.net.impl.ChromiumUrlRequestFactory;
import org.chromium.net.impl.CronetUrlRequest;
import org.chromium.net.impl.CronetUrlRequestContext;

/**
 * Utilities for Cronet testing
 */
@JNINamespace("cronet")
public class CronetTestUtil {
    private static final ConditionVariable sHostResolverBlock = new ConditionVariable();

    /**
     * Registers customized DNS mapping for testing host names used by test servers, namely:
     * <ul>
     * <li>{@link QuicTestServer#getServerHost}</li>
     * <li>{@link NativeTestServer#getSdchURL}</li>'s host
     * </ul>
     * @param cronetEngine {@link CronetEngine} that this mapping should apply to.
     * @param destination host to map to (e.g. 127.0.0.1)
     */
    public static void registerHostResolverProc(CronetEngine cronetEngine, String destination) {
        long contextAdapter =
                ((CronetUrlRequestContext) cronetEngine).getUrlRequestContextAdapter();
        nativeRegisterHostResolverProc(contextAdapter, false, destination);
        sHostResolverBlock.block();
        sHostResolverBlock.close();
    }

    /**
     * Registers customized DNS mapping for testing host names used by test servers.
     * @param requestFactory {@link HttpUrlRequestFactory} that this mapping should apply to.
     * @param destination host to map to (e.g. 127.0.0.1)
     */
    public static void registerHostResolverProc(
            HttpUrlRequestFactory requestFactory, String destination) {
        long contextAdapter = ((ChromiumUrlRequestFactory) requestFactory)
                                      .getRequestContext()
                                      .getUrlRequestContextAdapter();
        nativeRegisterHostResolverProc(contextAdapter, true, destination);
        sHostResolverBlock.block();
        sHostResolverBlock.close();
    }

    /**
     * Returns the value of load flags in |urlRequest|.
     * @param urlRequest is the UrlRequest object of interest.
     */
    public static int getLoadFlags(UrlRequest urlRequest) {
        return nativeGetLoadFlags(((CronetUrlRequest) urlRequest).getUrlRequestAdapterForTesting());
    }

    @CalledByNative
    private static void onHostResolverProcRegistered() {
        sHostResolverBlock.open();
    }

    private static native void nativeRegisterHostResolverProc(
            long contextAdapter, boolean isLegacyAPI, String destination);

    private static native int nativeGetLoadFlags(long urlRequest);
}
